### ISSUE TITLE
Web hook does not wait for completion of "mergeable" check

### ISSUE THREAD

**Issue by bilderbuchi** [go-gitea/gitea#19755]
### Description

Sometimes a webhook is triggered while the UI still displays 
> Merge conflict checking is in progress. Try again in few moments. 

after some action. This means that the content of the `mergeable` field is stale, so the receiver (in my case buildbot) ignores the PR event with 
> Received event 'pull_request' from gitea
> Gitea Pull Request ignored because it is not mergeable. (payload: action reopened, mergeable false)

A couple of seconds after the action, merge checking is completed and shows the PR as mergeable. 

I have observed this on 
* a `reopened` action, and on
* a `synchronized` action when the previous PR state had merge conflicts

which led me to the conclusion that the previous mergeable status is put in the payload.

I have observed this on a self-hosted Gitea 1.15, 1.16.something and now under 1.16.8.

### Expected behaviour
The webhook firing should be delayed until the merge conflict checking has completed. Otherwise, wrong `mergeable` status is reported in the payload.

The instance is not under my control, so I unfortunately cannot provide most of the debugging info below. :-(

### Gitea Version

1.16.8

### Can you reproduce the bug on the Gitea demo site?

did not attempt yet

### How are you running Gitea?

Self-hosted instance not under my control, details unknown.

**Comment by bilderbuchi** [comment]
Is there any more info you need on this bug report?

**Comment by r7l** [comment]
I can confirm this bug as it happened to me as well. The `mergeable` flag will even stay `false` regardless of the current state of the PR. Even when manually triggering the same webhook again.

**Comment by lunny** [comment]
> > Mergable checking may take very long.
> 
> OK. Does this have an impact on the propsed solution? I'd think it doesn't matter when exactly the webhook fires?
> 
> > The webhook firing should be delayed until the merge conflict checking has completed. Otherwise, wrong mergeable status is reported in the payload.

I don't know whether people would like the webhook is triggered fast or just wait the mergable detection. Maybe we can have some votes here.

**Comment by bilderbuchi** [comment]
Users: If you prefer that webhooks trigger more quickly, but can contain an _incorrect_ `mergeable` status, upvote this comment 👍

**Comment by bilderbuchi** [comment]
Users: If you prefer that the webhook payload always contains a _correct_ `mergeable` entry, even if it means that webhooks are delayed (by on the order of a couple of seconds?), upvote this comment 👍

**Comment by r7l** [comment]
Seems like it's just the 2 of us and i've pushed a PR to the library i am using in order to skip this check entirely. It doesn't work for me and only causes issues. I would like to make use of it but for that it would need to work properly.

What timeframe we are talking about here. I guess it's depending on the size of the project. Could it be seconds, minutes or even hours?

**Comment by bilderbuchi** [comment]
btw, related issue #25849

**Comment by lunny** [comment]
Or could we have a `mergable == checking` before the check finished.

**Comment by bilderbuchi** [comment]
> Or could we have a `mergable == checking` before the check finished.

...and then fire a second webhook as soon as the check is finished, potentially triggering a second CI build downstream? Is that not needlessly complicating any API handling logic of your consumers that want to avoid that?

It would be interesting to get a feeling for the answer to the above question: 
> What timeframe we are talking about here. I guess it's depending on the size of the project. Could it be seconds, minutes or even hours?

I have personally only ever seen O(seconds) (less than 5 surely, but haven't measured objectively) for completing the merge check, but I'm not maintaining the Linux kernel or other huge projects.
@lunny you seem to be concerned that 
> Mergable checking may take very long.

Do you have any data or experience to share here?

**Comment by lunny** [comment]
I would like to know how GH handle the problem especially for the big push.

**Comment by bilderbuchi** [comment]
> I would like to know how GH handle the problem especially for the big push.

Apparently a background job is involved. Here are some API docs that touch mergeability checks: 
https://docs.github.com/en/rest/guides/using-the-rest-api-to-interact-with-your-git-database?apiVersion=2022-11-28#checking-mergeability-of-pull-requests
https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#get-a-pull-request

However, those are about actively querying the PR status via API, not about the webhook timing, and if the webhook is delayed until the check is complete. 

As consumers of webhooks passively receive them, I'd think mandating them to actively trigger other API calls to get up-to-date information is weird/undesirable, as opposed to waiting some time for the check to complete. 
At least in my perspective, waiting O(seconds-minutes) longer for a webhook to fire would be a total non-issue, but modifying the CI runner (or any downstream consumer, really) that consumes those webhooks to implement additional logic to poll&re-check would be difficult or impossible :shrug: (and useless/ineffective anyway as long as #25849 is unfixed)

**Comment by lunny** [comment]
Maybe the attribute should have 3 states.

- `false`: cannot be merged
- `checking`: it's being checked, the consume can do an extra API to check later.
- `true`: can be merged.

This will not wait for the `checking` because most `CI` doesn't need this field.

**Comment by bilderbuchi** [comment]
See https://docs.github.com/en/graphql/reference/enums#mergeablestate

> because most CI doesn't need this field.

This assertion is surprising to me. Do most CI systems just start a build, only to fall down because the merge does not succeed? Or are you referring those that don't build the merge commit (as they should, arguably), but just the PR head?

**Comment by lunny** [comment]
Yes, I think Github encountered the same problem as us, so why there are 3 values for `MergeableState`.

> Values for MergeableState
> * CONFLICTING
> The pull request cannot be merged due to merge conflicts.
> * MERGEABLE
> The pull request can be merged.
> * UNKNOWN
> The mergeability of the pull request is still being calculated.

The PR head can have no conflict with the base branch(base commit) but have conflicts with the merge base. So I think even if it cannot be merged, the build may still run and succeed.

**Comment by lunny** [comment]
Added a test in https://github.com/go-gitea/gitea/pull/35337/commits/e433ff31d63aba6f2e142bd9884354872d0bfca0 to confirm creating pull request should get the right mergeable value.
